src/hotspot/share/opto/loopnode.cpp

```diff
@@ -3990,18 +4024,34 @@ void PhaseIdealLoop::replace_parallel_iv(IdealLoopTree *loop) {
       // variable differs from the trip counter by a loop-invariant
       // amount, the difference between their respective initial values.
       // It is scaled by the 'ratio_con'.
-      Node* ratio = _igvn.intcon(ratio_con);
+      Node* ratio = _igvn.integercon(ratio_con, stride_con2_bt);
       set_ctrl(ratio, C->root());
-      Node* ratio_init = new MulINode(init, ratio);
-      _igvn.register_new_node_with_optimizer(ratio_init, init);
+
+      Node* init_converted = init;
+      if (stride_con2_bt == T_LONG) {
+        init_converted = new ConvI2LNode(init);
+        _igvn.register_new_node_with_optimizer(init_converted, init);
+        set_early_ctrl(init_converted, false);
+      }
+
+      Node* ratio_init = MulNode::make(init_converted, ratio, stride_con2_bt);
+      _igvn.register_new_node_with_optimizer(ratio_init, init_converted);
       set_early_ctrl(ratio_init, false);
-      Node* diff = new SubINode(init2, ratio_init);
+      Node* diff = SubNode::make(init2, ratio_init, stride_con2_bt);
       _igvn.register_new_node_with_optimizer(diff, init2);
       set_early_ctrl(diff, false);
-      Node* ratio_idx = new MulINode(phi, ratio);
-      _igvn.register_new_node_with_optimizer(ratio_idx, phi);
+
+      Node* phi_converted = phi;
+      if (stride_con2_bt == T_LONG) {
+        phi_converted = new ConvI2LNode(phi);
+        _igvn.register_new_node_with_optimizer(phi_converted, phi);
+        set_early_ctrl(phi_converted, false);
+      }
+
+      Node* ratio_idx = MulNode::make(phi_converted, ratio, stride_con2_bt);
+      _igvn.register_new_node_with_optimizer(ratio_idx, phi_converted);
```

Comment:
This block has some code duplication. Could you refactor it somehow to make it more concise?